pkg/docker/config/config.go

```diff
@@ -0,0 +1,281 @@
+	if err != nil {
+		return "", "", err
+	}
+	return username, password, nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) error {
+	registry = DockerHostNameToRegistry(registry)
+	path, err := getPathToAuth()
+	if err != nil {
+		return err
+	}
+	return modifyJSON(registry, "", "", path, deleteAuth)
+}
+
+// DockerHostNameToRegistry converts the docker hostname to the docker registry
+func DockerHostNameToRegistry(registry string) string {
+	if registry == dockerHostname {
+		return dockerRegistry
+	}
+	return registry
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0700); err != nil {
+			return "", errors.Wrapf(err, "error creating directory %q", dir)
+		}
+	}
+	return filepath.Join(dir, authCfgFileName), nil
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file and returns it
+// or returns an empty dockerConfigFile data structure if auth.json is empty
+func readJSONFile(path string) (dockerConfigFile, error) {
+	var auths dockerConfigFile
+
+	raw, err := ioutil.ReadFile(path)
+	if os.IsNotExist(err) {
+		auths.AuthConfigs = map[string]dockerAuthConfig{}
+		return auths, nil
+	}
+
+	if err = json.Unmarshal(raw, &auths); err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
+	}
+
+	return auths, nil
+}
+
+// addAuthToJSON either adds a new registry url with its encoded user:pass
+// or modifies an existing registry url
+func modifyJSON(registry, username, password, path, action string) error {
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+
+	switch action {
+	case addAuth:
+		if len(auths.CredHelpers) > 0 {
+			if err = setAuthToCredHelper(registry, username, password); err != nil {
+				return err
+			}
+			break
+		}
+
+		creds := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
+		newCreds := dockerAuthConfig{Auth: creds}
+		auths.AuthConfigs[registry] = newCreds
+	case deleteAuth:
+		// First try cred helpers.
+		if ch, exists := auths.CredHelpers[registry]; exists {
+			if err = deleteAuthFromCredHelper(ch, registry); err != nil {
+				return err
+			}
+			break
+		}
+
+		if _, ok := auths.AuthConfigs[registry]; ok {
+			delete(auths.AuthConfigs, registry)
+		} else if _, ok := auths.AuthConfigs[normalizeRegistry(registry)]; ok {
+			delete(auths.AuthConfigs, normalizeRegistry(registry))
+		} else {
+			return ErrNotLoggedIn
+		}
```

Comment:
To reduce duplication of code, I change `addAuthToJSON` to `modifyJSON` where an entry is either added or deleted based on `action`. This function looks like a lot to take in, so I was thinking of writing separate function like `addCred` and `deleteCreds` and call those in here.
What is your opinion @mtrmac ?

Reply:
That would indeed be better than switching on a string. The `addCred`/`deleteCreds` could be placed close to `SetAuthentication`/`RemoveAuthentication` to make it easier to see how they work.

Even better, I think, would be 
```go
func modifyJSON(path string, editor func(auths *dockerConfigFile) error) (updated bool, error) {
   auths := …
   updated, err := editor(auths)
   if err != nil …
   if updated { /* Marshal + WriteFile */ }
}
```
and
```go
func SetAuthentication(…) error {
 …
 return modifyJSON(path, func(auths *dockerConfigFile) (bool, error) {
   /* the code from `case addAuth */
 })
}
```

then `modifyJSON` does not need to know anything about registries, usernames, passwords: e.g. `RemoveAuthentication` will avoid the slightly awkward `"", ""` values for user/password.

Reply:
… actually, with the `updated` boolean, the `func()…` approach still introduces some extra complexity which might not be worth it; perhaps there just should be `readJSON`/`writeJSON` helpers called by the `Set`/`Remove` methods, and no fancy lambdas.

Choose whichever you like best.